### ISSUE TITLE
lldb RISC-V support

### ISSUE THREAD

**Issue by kasperk81** [llvm/llvm-project#55383]
On RISC-V, gdb is not always a suitable option to debug programs, especially for folks who use lldb plugins on other architectures. Without the working lldb, debugging code can be quite challenging.

[D62732](https://reviews.llvm.org/D62732) is in review for over three years and there is no indication whether it is moving forward or abandoned. I realize that it is a matter of prioritization, but if someone else in the community can help with testing on riscv64/32, please do so. Or better yet, if you are familiar with lldb code-base, resubmit the patch.

**Comment by kasperk81** [comment]
@SEmmmer i saw your implementation in main branch. can you tell what is the status of lldb support on risc-v, is it completed (with system-v, plugins etc.)?

**Comment by DavidSpickett** [comment]
Someone asked about this on Discord, so clearly we need a bit of an update here.

The referenced Phab review did later land as https://github.com/llvm/llvm-project/commit/847de9c332775d1841fec9fea5cb5c41592a4c8f.

Recently we got corefile support for RISC-V 64: https://github.com/llvm/llvm-project/commit/d3a9043ec2ee7ea278be4f3c86823512e44d01bf

Before that, Emmer added all the register information and a whole bunch of instructions to the emulation plugin. I think this was needed due to a lack of hardware single step (a standardized one at least).

I've been told the basics work when connecting to qemu at least, but I don't know about a native RISC-V system and there's no publuc build bot for either configuration.

If anyone does try lldb and has problems, please raise issues for them, it helps a lot just to know they exist.

@tedwoodward you seem to have been in this area last year, do you know what the state is?

**Comment by tedwoodward** [comment]
Earlier this year, Qualcomm released a toolchain based on LLVM 18.0 that included LLDB. It works with RISC-V 32 and 64, tested with Linux user space QEMU.

Our team is working on getting the LLDB test suite running for RISC-V 32 using QEMU, and plan to upstream support when it's in good shape. I'd like to run a bot in the llvm farm, but we haven't had that discussion internally. If not us, I'd like the tests to be set up so anyone could get RISC-V QEMU and run LLDB tests with it.

I don't think we've got any debug on hardware yet. Maybe when there is more hardware, and RISC-V Linux is more mature we could port lldb-server to it, and run remote tests on hardware (or system QEMU). Or JTAG with OpenOCD.

**Comment by DavidSpickett** [comment]
Significant PRs since Ted's update in July:
https://github.com/llvm/llvm-project/pull/99039
https://github.com/llvm/llvm-project/pull/99043
https://github.com/llvm/llvm-project/pull/90075
https://github.com/llvm/llvm-project/pull/93297
https://github.com/llvm/llvm-project/pull/79990
https://github.com/llvm/llvm-project/pull/104547

In summary:
* A bunch of API tests fixed.
* Fixed breakpointing an unknown instruction.
* Core file debugging for rv64, including gpr only and gpr + fpr.
* Improved connections to non lldb-server remote debug servers on risc-v.

**Comment by DavidSpickett** [comment]
Expressions can now make function calls: https://github.com/llvm/llvm-project/pull/99336

**Comment by ncopa** [comment]
im currently working on llvm19 for alpine.
We only need https://github.com/llvm/llvm-project/pull/104547.patch for 19 right?

**Comment by sylvestre** [comment]
They are in the correct order?

**Comment by sylvestre** [comment]
ok, could you please get these patches in the next 19 dot releases?
once it is done, i will see how to apply it to 18 deb packages
thanks

**Comment by tru** [comment]
> @tru @DavidSpickett,
> 
> 
> 
> Would it be feasible to backport the following LLDB patches for RISC-V to v19 branch, or do they not meet the criteria for backporting?
> 
> 
> 
> Justification: RISC-V is an emerging architecture, and LLDB's support for it is still maturing. These patches help bring RISC-V closer to feature parity with other architectures currently supported by LLVM.
> 
> 
> 
> Risk: The risk is low, as LLDB’s RISC-V support is relatively new, and the associated functionality has not been previously operational.
> 
> 
> 
> for chronological cherry-picking, this is the order: https://github.com/llvm/llvm-project/issues/55383#issuecomment-2405385018 (+/- potential conflicts 🙂)


It would definitely be way outside what we normally consider for back-porting. Adding new features post the first release doesn't seem sensible to me, even if the risk for other platforms are low in theory, it is a risk at this point.

**Comment by DavidSpickett** [comment]
I agree. If these patches were fixing issues in features already supported on RISC-V, might be a different story.

Which means anyone interested in RISC-V support has a few months to make LLDB 20 as good as possible. Please raise issues for any test failures and missing features you or other users find. Use builds from the `main` branch if possible, so problems can be fixed before they make it into a release branch.

Also there is a talk at the upcoming [dev meeting](https://discourse.llvm.org/t/announcing-the-2024-llvm-developers-meeting-program/81108) on RISC-V support "Hey, do you want a RISC-V debugger?”, so you might get some good news before the next release.
